tests/test_init.py

```diff
@@ -26,3 +26,208 @@ def test_systemd_present(container):
+    # extracts a timestamp with a given prefix and converts it to seconds
+    def extract_time(stdout, prefix):
+        i = stdout.find(prefix)
+        if i < 0:
+            raise ValueError("prefix not found")
+        t = (
+            stdout[i + len(prefix) + 1 :].strip().split(" ")[0]
+        )  # take the next word
+        if t.endswith("ms"):
+            print("word:", t[:-2])
+            return float(t[:-2]) / 1000.0
+        elif t.endswith("s"):
+            print("word:", t[:-1])
+            return float(t[:-1])
+        else:
+            raise ValueError("time unit not recognized")
+
+    time = container.connection.run_expect([0], "systemd-analyze time")
+    startup = extract_time(time.stdout, "Startup finished in ")
+    assert startup < 5, "Startup threshold exceeded"
+    target = extract_time(time.stdout, ".target reached after ")
+    assert target < 5, "Reaching systemd target threshold exceeded"
+
+
+@pytest.mark.parametrize("container", [INIT_CONTAINER], indirect=True)
+@pytest.mark.skipif(
+    DOCKER_SELECTED
+    and (int(LOCALHOST.package("systemd").version.split(".")[0]) >= 248),
+    reason="Running systemd in docker is broken as of systemd 248, see https://github.com/moby/moby/issues/42275",
+)
+def test_systemd_nofailed_units(container):
+    """
+    Check that there are not failed systemd units present
+    """
+    output = container.connection.run_expect(
+        [0], "systemctl list-units --state=failed"
+    )
+    assert (
+        "0 loaded units listed" in output.stdout
+    ), "failed systemd units detected"
+
+
+@pytest.mark.parametrize("container", [INIT_CONTAINER], indirect=True)
+@pytest.mark.skipif(
+    DOCKER_SELECTED
+    and (int(LOCALHOST.package("systemd").version.split(".")[0]) >= 248),
+    reason="Running systemd in docker is broken as of systemd 248, see https://github.com/moby/moby/issues/42275",
+)
+def test_systemd_nofailed_units(container):
+    """
+    Check that there are not failed systemd units present
+    """
+    output = container.connection.run_expect(
+        [0], "systemctl list-units --state=failed"
+    )
+    assert (
+        "0 loaded units listed" in output.stdout
+    ), "failed systemd units detected"
+
+
+@pytest.mark.parametrize("container", [INIT_CONTAINER], indirect=True)
+@pytest.mark.skipif(
+    DOCKER_SELECTED
+    and (int(LOCALHOST.package("systemd").version.split(".")[0]) >= 248),
+    reason="Running systemd in docker is broken as of systemd 248, see https://github.com/moby/moby/issues/42275",
+)
+def test_systemd_detect_virt(container, host, container_runtime):
+    """
+    Check that systemd-detect-virt correctly detects the used container runtime
+    """
+    output = container.connection.run_expect([0], "systemd-detect-virt")
+    runtime = container_runtime.runner_binary
+    assert (
+        runtime in output.stdout
+    ), f"systemd-detect-virt failed to detect {runtime}"
```

Comment:
```suggestion
    output = container.connection.run_expect([0], "systemd-detect-virt").stdout
    runtime = container_runtime.runner_binary
    assert (
        runtime in output
    ), "systemd-detect-virt failed to detect {runtime}"
```

Reply:
Yep, `output` suggests that this is the better handling.